### ISSUE TITLE
Confirm EOA `/wallet`

### ISSUE THREAD

**Issue by pavlovcik** [ubiquity/ubiquibot#410]
I'm not sure of the best strategy but to make things more noob proof we should try and verify that the wallet to register is an EOA and not an exchange deposit address (because they won't be able to use it to execute the claim transaction)  

1. perhaps there is an open source tool or framework to identify that the attempted wallet to be registered is a CEX deposit address
2. We could check if there is any gas in the wallet. If not then throw a warning explaining that the wallet must be a self custody wallet (e.g. metamask) for claims to work.

**Comment by web4er** [comment]
Another strategy is to get the bounty hunter to sign some message using metamask. This confirms that the wallet is EOA, and not from a CEX. 

There are two ways to do this.

1. When user changes wallet using ubiquitbot, it also provides a signed message with it. ubiquitbot  checks its validity and acts accordingly. This will require the bounty hunter to sign message manually.
2. Give bounty hunter a link to dApp or somewhere there is a wallet connection, and get some message signed.

**Comment by pavlovcik** [comment]
@web4er that's a cool idea! Here's an updated flow:

1. user invokes `/wallet` command with their address as the argument (exactly as we do currently.)
2. the bot posts a link to pay.ubq.fi/verify?user=pavlovcik&wallet=0x4007CE2083c7F3E18097aeB3A39bb8eC149a341d
3. the dApp parses the URL parameters and then asks for a signature
4. upon success there must be some type of webhook interface to the bot (@0xcodercrane @rndquu I would like your input here)
5. The bot posts a wallet registration success comment on the issue.

**Comment by rndquu** [comment]
We could enforce bounty hunters to register public keys instead of wallet addresses

So if bounty hunter knows a public key then they should also know the private key. CEX exchanges don't share such info.

On the bot's side we could derive address from the public key

This is not really user friendly though so perhaps we will have to create a new web page "get public key from authorized wallet address in metamask"

Example (register address):
```
/wallet 0xfF02FE4e3FD16DAA1C6A95ebDB8e1d62C6cE71AF
```

Example (register public key):
```
/wallet 0x038cfb8c0ad2f3c6f603f84e29283bd09fc29fd751f3318c5b69e98b241956ecfd
```

**Comment by pavlovcik** [comment]
> We could enforce bounty hunters to register public keys instead of wallet addresses

I definitely like the idea of less steps, but as you said, it's not very user-friendly. I wonder if there is a simple way to derive the public key without navigating to a new tool. Otherwise, the signature approach might be more intuitive for most hunters.

**Comment by pavlovcik** [comment]
I think step 2 and 3 are over complicating it. There isn't an incentive for people to lie about their wallet address.

**Comment by pavlovcik** [comment]
> the bot posts a link to pay.ubq.fi/verify?wallet=0x4007CE2083c7F3E18097aeB3A39bb8eC149a341d&issue=100

`pay.ubq.fi/verify?wallet=0x4007CE2083c7F3E18097aeB3A39bb8eC149a341d&owner=ubiquity&repository=ubiquibot&issue=100`

**Comment by web4er** [comment]
> > @web4er that's a cool idea! Here's an updated flow:
> > 
> > 1. user invokes `/wallet` command with their address as the argument (exactly as we do currently.)
> > 2. the bot posts a link to `pay.ubq.fi/verify?user=pavlovcik&wallet=0x4007CE2083c7F3E18097aeB3A39bb8eC149a341d`
> > 3. the dApp parses the URL parameters and then asks for a signature
> > 4. upon success there must be some type of webhook interface to the bot (@0xcodercrane @rndquu I would like your input here)
> > 5. The bot posts a wallet registration success comment on the issue.
> 
> Yes, it could work
> 
>     1. user invokes `/wallet` command with their address as the argument (exactly as we do currently.)
> 
>     2. the bot posts a link to `pay.ubq.fi/verify?wallet=0x4007CE2083c7F3E18097aeB3A39bb8eC149a341d&issue=100`
> 
>     3. user authenticates in github via OAuth
> 
>     4. user signs tx data
> 
>     5. user sends tx data to the bot's backend
> 
>     6. on the backend the bot verifies that the provided wallet equals to the one derived from the signature
> 
>     7. the bot posts a wallet registration success comment on the issue
> 
> 
> My only concern is step 5 because I'm not sure that it is possible to send a valid request with [user's token](https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/authenticating-with-a-github-app-on-behalf-of-a-user)

I think this is a lot of back and forth communication. Too complex relative to the benefit it is providing. 

Here is how I will simplify it. 

1. Once a user sets their wallet, they get a response from the bot to sign a message like "Contribution to ubiquity by @web4er". They can use any tool to sign it, myetherwallet.com, or etherscan https://etherscan.io/verifiedSignatures to sign it,  and reply with signature.
2. The bot checks for signature validity and responds accordingly. 


The same thing can be done if we use public key, but I couldn't find any trusted tool to extract public key.

**Comment by pavlovcik** [comment]
### Full Example

Address
```
0x4007ce2083c7f3e18097aeb3a39bb8ec149a341d
```


Message
> Contribution to Ubiquity by @pavlovcik

Signature Hash
``` 
0xcdf0936ec5a40c925dec7db9b6f2f9f6b76839da10abff412371b11cebf022a72e66776803e5dce178dbf63d9381cc535a5507dc90558498f9d34ee30f9622101b
```

**Comment by rndquu** [comment]
Isn't it better to make it a one liner? I mean bounty hunter would need to provide a wallet address + sign the provided wallet address:
```
/wallet 0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266 0x0830f316c982a7fd4ff050c8fdc1212a8fd92f6bb42b2337b839f2b4e156f05a359ef8f4acd0b57cdedec7874a865ee07076ab2c81dc9f9de28ced55228587f81c
```

**Comment by pavlovcik** [comment]
I think that one liner sounds like a plan! Also with regular expressions I think that its easy to identify the intent behind every argument regardless of the order. 

We just need to communicate the string that they need to hash somehow. So the flow could possibly be the following:

- A user writes `/wallet 0x4007CE2083c7F3E18097aeB3A39bb8eC149a341d`
- The bot renders an error comment 
  - "Use [Etherscan](https://etherscan.io/verifiedSignatures) to sign `ubiquity/ubiquibot/issues/410` and then register your wallet by appending the Signature Hash to the end of the command `/wallet 0x4007CE2083c7F3E18097aeB3A39bb8eC149a341d 0x0830f316c982a7fd4ff050c8fdc1212a8fd92f6bb42b2337b839f2b4e156f05a359ef8f4acd0b57cdedec7874a865ee07076ab2c81dc9f9de28ced55228587f81c`"

Unless there is a way to handle this flow more efficiently?

**Comment by rndquu** [comment]
To sum up:
1. Bounty hunter should be able to register a wallet with a signed message in any order. Examples:
```
/wallet 0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266 0x0830f316c982a7fd4ff050c8fdc1212a8fd92f6bb42b2337b839f2b4e156f05a359ef8f4acd0b57cdedec7874a865ee07076ab2c81dc9f9de28ced55228587f81c

/wallet 0x0830f316c982a7fd4ff050c8fdc1212a8fd92f6bb42b2337b839f2b4e156f05a359ef8f4acd0b57cdedec7874a865ee07076ab2c81dc9f9de28ced55228587f81c 0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266
```
2. If bounty hunter posts an invalid `wallet` command then the bot should respond with error:
  - "[Use Etherscan](https://etherscan.io/verifiedSignatures) to sign `ubiquity/ubiquibot/410` and then register your wallet by appending the Signature Hash to the end of the command `/wallet 0x4007CE2083c7F3E18097aeB3A39bb8eC149a341d 0x0830f316c982a7fd4ff050c8fdc1212a8fd92f6bb42b2337b839f2b4e156f05a359ef8f4acd0b57cdedec7874a865ee07076ab2c81dc9f9de28ced55228587f81c`"
3. We could sign either bounty hunter's wallet string either issue URL `ubiquity/ubiquibot/410`. Anyway in the end the bot should check that the message signer equals to the provided wallet address.

**Comment by web4er** [comment]
This is coming together very nicely. 

> We could sign either bounty hunter's wallet string either issue URL ubiquity/ubiquibot/410. 

I think signing the wallet itself has a little signature replay attack potential. I will not consider it as my first choice.

On the other hand, does signing a dynamic message like `ubiquity/ubiquibot/410`  has a benefit? Signing a static message like `ubiquity/ubiquibot` has some less significant benefits. 
1. Easy to explain in help.
2. Hunter will have to sign the message only once and he can reuse the signed message hash.
3. Dynamic message can give the impression that hunter can use different wallets on different issues because of prepended issue number. While the static message doesn't. 

Here is how the help will look. 

```
.
.
.
- /wallet: <WALLET_ADDRESS | ENS_NAME> <SIGNED_MESSAGE_HASH>: Register the hunter's wallet address. 
  The message that you have to sign is ubiquity/ubiquibot
  You can generate SIGNED_MESSAGE_HASH on etherscan https://etherscan.io/verifiedSignatures
  ex1: /wallet 0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266 0x75329f883590507e581cd6dfca62680b6cd12e1f1665db8097f9e642ed70025146b5cf9f777dde90c4a9cbd41500a6bf76bc394fd0b0cae2aab09f7a6f30e3b31b
  ex2: /wallet vitalik.eth 0x75329f883590507e581cd6dfca62680b6cd12e1f1665db8097f9e642ed70025146b5cf9f777dde90c4a9cbd41500a6bf76bc394fd0b0cae2aab09f7a6f30e3b31b
```

**Comment by web4er** [comment]
> To sum up:
> 
>     1. Bounty hunter should be able to register a wallet with a signed message in any order. Examples:
> 
> 
> ```
> /wallet 0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266 0x0830f316c982a7fd4ff050c8fdc1212a8fd92f6bb42b2337b839f2b4e156f05a359ef8f4acd0b57cdedec7874a865ee07076ab2c81dc9f9de28ced55228587f81c
> 
> /wallet 0x0830f316c982a7fd4ff050c8fdc1212a8fd92f6bb42b2337b839f2b4e156f05a359ef8f4acd0b57cdedec7874a865ee07076ab2c81dc9f9de28ced55228587f81c 0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266
> ```
> 
>     2. If bounty hunter posts an invalid `wallet` command then the bot should respond with error:
> 
> 
>     * "[Use Etherscan](https://etherscan.io/verifiedSignatures) to sign `ubiquity/ubiquibot/410` and then register your wallet by appending the Signature Hash to the end of the command `/wallet 0x4007CE2083c7F3E18097aeB3A39bb8eC149a341d 0x0830f316c982a7fd4ff050c8fdc1212a8fd92f6bb42b2337b839f2b4e156f05a359ef8f4acd0b57cdedec7874a865ee07076ab2c81dc9f9de28ced55228587f81c`"
> 
> 
>     3. We could sign either bounty hunter's wallet string either issue URL `ubiquity/ubiquibot/410`. Anyway in the end the bot should check that the message signer equals to the provided wallet address.

This looks very close to the final requirements. I would love to work on this on the weekend.

**Comment by web4er** [comment]
I have worked on this over the weekend. Here are tests of how it is working https://github.com/web4er/obiquibot-tests/issues/3

Feel free to suggest any improvements. Looking forward to the bounty and opening my first pull request.

**Comment by pavlovcik** [comment]
> I think signing the wallet itself has a little signature replay attack potential. I will not consider it as my first choice.

I'm not sure if signature replay is a valid exploit vector because wouldn't the attacker also need to have access to the user's GitHub account? Perhaps I am not realizing the full attack surface area.

> On the other hand, does signing a dynamic message like ubiquity/ubiquibot/410 has a benefit? Signing a static message like ubiquity/ubiquibot has some less significant benefits.

When I picked `ubiquity/ubiquibot/410` it was chosen because it feels more intentional than what the previous proposal suggested. I didn't fully think through if it is the optimal choice. 

Ultimately it seems like signing a static message makes the most sense to make it simple to explain. I would like to choose a static string that doesn't feel completely random!

**Comment by pavlovcik** [comment]
I think it's worth adding a config for this to enable/disable this security feature. Otherwise, to fallback to the old behavior of just setting any address (it was a lot simpler!) 

We can do this in a separate bounty. I imagine that asking non-technical bounty hunters to provide a signature hash could be somewhat overwhelming for them.

**Comment by web4er** [comment]
I have submitted a draft pull request. My code is ready for review, but I am unable to understand the QA bot. It is responding without my changes. You can see https://github.com/web4er/ubiquibot/issues/11#issuecomment-1622607074  that my bot is not accepting wallet without a signature, but QA bot is still accepting. What am I missing?

**Comment by web4er** [comment]
> I'm not sure if signature replay is a valid exploit vector because wouldn't the attacker also need to have access to the user's GitHub account? Perhaps I am not understanding the full attack strategy.

The signature posted here is public. It can be reused on some other Dapp that the user uses. If the Dapp provides some access/benefit authorized by a signed wallet address, it will be a problem.  

> That looks great! Awesome tests. Why not open the draft pull request with its source code?

Thank you, I have opened a draft pull request.

**Comment by pavlovcik** [comment]
> My code is ready for review, but I am unable to understand the QA bot. It is responding without my changes. You can see web4er#11 (comment) that my bot is not accepting wallet without a signature, but QA bot is still accepting. What am I missing?

@0xcodercrane @rndquu how does the QA bot (GitHub actions runtime on pull requests) read the newly modified code? Can we fix this?

**Comment by rndquu** [comment]
> > My code is ready for review, but I am unable to understand the QA bot. It is responding without my changes. You can see [web4er#11](https://github.com/web4er/ubiquibot/issues/11) (comment) that my bot is not accepting wallet without a signature, but QA bot is still accepting. What am I missing?
> 
> @0xcodercrane @rndquu how does the QA bot (GitHub actions runtime on pull requests) read the newly modified code? Can we fix this?

Perhaps when the QA bot is invoked from an issue (not from a PR) then the `development` branch is checked out so if the updated code resides in some other branch then the none-modified `development` branch is invoked instead of the feature branch

@web4er try to check how it works in the PR (which contains the updated code) instead of an issue

**Comment by 0xcodercrane** [comment]
> > My code is ready for review, but I am unable to understand the QA bot. It is responding without my changes. You can see [web4er#11](https://github.com/web4er/ubiquibot/issues/11) (comment) that my bot is not accepting wallet without a signature, but QA bot is still accepting. What am I missing?
> 
> @0xcodercrane @rndquu how does the QA bot (GitHub actions runtime on pull requests) read the newly modified code? Can we fix this?

If you run the QA bot with the changes, the changes need to be merged into the `development` branch first because the workflow in the default branch: `development` is triggered for the issue related events.

**Comment by web4er** [comment]
> If you run the QA bot with the changes, the changes need to be merged into the development branch first because the workflow in the default branch: development is triggered for the issue-related events. 

Ok, I have merged changes to `development` branch and it started to work and responded to some commands. But now it is not responding to the commands after https://github.com/web4er/ubiquibot/issues/11#issuecomment-1623520409 and it has been an hour.  @0xcodercrane

**Comment by web4er** [comment]
The QA bot replied after 5 hours. The QA is complete but not as orderly as I wanted it to be. I tried again to QA but the bot has again disappeared now. 

I am going to submit the PR for review now.

**Comment by pavlovcik** [comment]
> But now it is not responding to the commands after [web4er#11 (comment)](https://github.com/web4er/ubiquibot/issues/11#issuecomment-1623520409) and it has been 2 hours. @0xcodercrane

That's wild. What could be causing the insane lag? 

@0xcodercrane @rndquu check how long some of these are taking to run: https://github.com/web4er/ubiquibot/actions

**Comment by 0xcodercrane** [comment]
I had the same feeling that the github workflow took so long even not completed. It made me to rerun the workflow. I wanna say it is kind of a GitHub itself issue.

**Comment by 0xcodercrane** [comment]
@web4er the quick way to verify it is to cancel the workflow and rerun and see how it goes

**Comment by pavlovcik** [comment]
> I had the same feeling that the github workflow took so long even not completed. It made me to rerun the workflow. I wanna say it is kind of a GitHub itself issue.

I don't think it's valid to point the blame at GitHub and not investigate. 

We should debug the reason why this hangs. I think we could make a shorter timeout for the time being.

**Comment by rndquu** [comment]
Let's wait for more such cases because some other workflows (`Build`, `Kebab case`) also took too long to finish which means that the reason could be a github

<img width="1512" alt="Screenshot 2023-07-07 at 12 37 26" src="https://github.com/ubiquity/ubiquibot/assets/119500907/2d8d8982-f4c1-4572-8220-aeec93e302a0">
